File: frontend/src/lib/training.ts
const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/training`;

async function req<T>(path: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...opts,
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? r.statusText);
  }
  if (r.status === 204) return undefined as T;
  return r.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type SkillCategory = "technical" | "soft" | "safety" | "compliance" | "management";
export type SkillLevel = "basic" | "intermediate" | "advanced" | "expert";
export type TrainingType = "internal" | "external" | "online" | "classroom" | "on_job";
export type SessionStatus = "planned" | "in_progress" | "completed" | "cancelled";
export type AssignmentStatus = "assigned" | "in_progress" | "completed" | "overdue" | "waived";
export type CertificationStatus = "valid" | "expiring" | "expired" | "revoked";
export type TRAIAgentType = "skill_gap_advisor" | "effectiveness_analyzer" | "compliance_risk";
export type TRAIRecStatus = "pending" | "acknowledged" | "actioned" | "dismissed";

export interface Skill {
  skill_id: string;
  skill_code: string;
  skill_name: string;
  category: SkillCategory;
  description: string | null;
  active_flag: boolean;
  created_at: string;
}

export interface EmployeeSkillRow {
  employee_skill_id: string;
  employee_id: string;
  employee_name: string | null;
  department_id: string | null;
  skill_id: string;
  skill_name: string;
  skill_category: SkillCategory;
  current_level: SkillLevel | null;
  required_level: SkillLevel | null;
  has_gap: boolean;
  last_assessed_date: string | null;
  notes: string | null;
  created_at: string;
}

export interface TrainingProgram {
  training_id: string;
  training_code: string;
  training_name: string;
  training_type: TrainingType;
  category: SkillCategory;
  duration_hours: number | null;
  provider: string | null;
  cost: number | null;
  mandatory_flag: boolean;
  validity_period_days: number | null;
  active_flag: boolean;
  notes: string | null;
  created_at: string;
}

export interface TrainingSession {
  session_id: string;
  training_id: string;
  training_name: string | null;
  session_date: string;
  end_date: string | null;
  location: string | null;
  trainer_id: string | null;
  trainer_name: string | null;
  capacity: number | null;
  enrolled_count: number;
  status: SessionStatus;
  notes: string | null;
  created_at: string;
}

export interface TrainingAssignment {
  assignment_id: string;
  employee_id: string;
  employee_name: string | null;
  department_id: string | null;
  training_id: string;
  training_name: string | null;
  session_id: string | null;
  assigned_by: string | null;
  due_date: string | null;
  completion_date: string | null;
  status: AssignmentStatus;
  score: number | null;
  notes: string | null;
  created_at: string;
}

export interface CertificationRecord {
  certification_id: string;
  employee_id: string;
  employee_name: string | null;
  training_id: string | null;
  certificate_name: string;
  certificate_no: string | null;
  issuer: string | null;
  issue_date: string;
  expiry_date: string | null;
  status: CertificationStatus;
  days_to_expiry: number | null;
  document_url: string | null;
  notes: string | null;
  created_at: string;
}

export interface TrainingFeedback {
  feedback_id: string;
  employee_id: string;
  employee_name: string | null;
  training_id: string;
  session_id: string | null;
  rating: number;
  content_rating: number | null;
  trainer_rating: number | null;
  relevance_rating: number | null;
  comments: string | null;
  created_at: string;
}

export interface TrainingDashboard {
  total_programs: number;
  active_programs: number;
  total_sessions: number;
  upcoming_sessions: number;
  total_assignments: number;
  completed_assignments: number;
  overdue_assignments: number;
  total_certifications: number;
  expiring_soon: number;
  expired_certifications: number;
  total_skill_profiles: number;
  skill_gaps: number;
  avg_feedback_rating: number | null;
}

export interface TRAIRec {
  rec_id: string;
  agent_type: TRAIAgentType;
  status: TRAIRecStatus;
  title: string;
  body: string;
  employee_id: string | null;
  department_id: string | null;
  training_id: string | null;
  certification_id: string | null;
  score: number | null;
  actioned_by: string | null;
  actioned_at: string | null;
  created_at: string;
}

// ── Color/Label Maps ──────────────────────────────────────────────────────────

export const CATEGORY_COLOR: Record<SkillCategory, string> = {
  technical: "bg-blue-100 text-blue-700",
  soft: "bg-purple-100 text-purple-700",
  safety: "bg-red-100 text-red-700",
  compliance: "bg-orange-100 text-orange-700",
  management: "bg-indigo-100 text-indigo-700",
};

export const LEVEL_COLOR: Record<SkillLevel, string> = {
  basic: "bg-gray-100 text-gray-600",
  intermediate: "bg-yellow-100 text-yellow-700",
  advanced: "bg-blue-100 text-blue-700",
  expert: "bg-green-100 text-green-700",
};

export const LEVEL_ORDER: Record<SkillLevel, number> = {
  basic: 1, intermediate: 2, advanced: 3, expert: 4,
};

export const SESSION_STATUS_COLOR: Record<SessionStatus, string> = {
  planned: "bg-blue-100 text-blue-700",
  in_progress: "bg-yellow-100 text-yellow-700",
  completed: "bg-green-100 text-green-700",
  cancelled: "bg-gray-100 text-gray-500",
};

export const ASSIGNMENT_STATUS_COLOR: Record<AssignmentStatus, string> = {
  assigned: "bg-blue-100 text-blue-700",
  in_progress: "bg-yellow-100 text-yellow-700",
  completed: "bg-green-100 text-green-700",
  overdue: "bg-red-100 text-red-700",
  waived: "bg-gray-100 text-gray-500",
};

export const CERT_STATUS_COLOR: Record<CertificationStatus, string> = {
  valid: "bg-green-100 text-green-700",
  expiring: "bg-yellow-100 text-yellow-800",
  expired: "bg-red-100 text-red-700",
  revoked: "bg-gray-100 text-gray-500",
};

export const TRAINING_TYPE_LABEL: Record<TrainingType, string> = {
  internal: "Internal",
  external: "External",
  online: "Online",
  classroom: "Classroom",
  on_job: "On-the-Job",
};

// ── API Client ────────────────────────────────────────────────────────────────

export const trainingApi = {
  getDashboard: () => req<TrainingDashboard>("/dashboard"),

  // Skills
  createSkill: (d: object) => req<Skill>("/skills", { method: "POST", body: JSON.stringify(d) }),
  listSkills: (category?: string, active?: boolean) => {
    const q = new URLSearchParams();
    if (category) q.set("category", category);
    if (active) q.set("active_only", "true");
    return req<Skill[]>(`/skills?${q}`);
  },
  getSkill: (id: string) => req<Skill>(`/skills/${id}`),
  updateSkill: (id: string, d: object) => req<Skill>(`/skills/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  seedSkills: () => req<{ created: number }>("/skills/seed-defaults", { method: "POST" }),

  // Skill Profiles
  upsertSkillProfile: (d: object) => req<EmployeeSkillRow>("/skill-profiles", { method: "POST", body: JSON.stringify(d) }),
  updateSkillProfile: (id: string, d: object) =>
    req<EmployeeSkillRow>(`/skill-profiles/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  getSkillMatrix: (employee_id?: string, department_id?: string) => {
    const q = new URLSearchParams();
    if (employee_id) q.set("employee_id", employee_id);
    if (department_id) q.set("department_id", department_id);
    return req<EmployeeSkillRow[]>(`/skill-matrix?${q}`);
  },
  getSkillGapReport: (department_id?: string) =>
    req<EmployeeSkillRow[]>(`/skill-gap-report${department_id ? `?department_id=${department_id}` : ""}`),

  // Programs
  createProgram: (d: object) => req<TrainingProgram>("/programs", { method: "POST", body: JSON.stringify(d) }),
  listPrograms: (params?: { category?: string; mandatory_only?: boolean; active_only?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.category) q.set("category", params.category);
    if (params?.mandatory_only) q.set("mandatory_only", "true");
    if (params?.active_only) q.set("active_only", "true");
    return req<TrainingProgram[]>(`/programs?${q}`);
  },
  getProgram: (id: string) => req<TrainingProgram>(`/programs/${id}`),
  updateProgram: (id: string, d: object) =>
    req<TrainingProgram>(`/programs/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  getProgramEffectiveness: (id: string) => req<object>(`/programs/${id}/effectiveness`),

  // Sessions
  createSession: (d: object) => req<TrainingSession>("/sessions", { method: "POST", body: JSON.stringify(d) }),
  listSessions: (params?: { training_id?: string; status?: string; upcoming_only?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.training_id) q.set("training_id", params.training_id);
    if (params?.status) q.set("status", params.status);
    if (params?.upcoming_only) q.set("upcoming_only", "true");
    return req<TrainingSession[]>(`/sessions?${q}`);
  },
  getSession: (id: string) => req<TrainingSession>(`/sessions/${id}`),
  updateSession: (id: string, d: object) =>
    req<TrainingSession>(`/sessions/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  completeSession: (id: string) =>
    req<TrainingSession>(`/sessions/${id}/complete`, { method: "POST" }),

  // Assignments
  createAssignment: (d: object) =>
    req<TrainingAssignment>("/assignments", { method: "POST", body: JSON.stringify(d) }),
  listAssignments: (params?: { employee_id?: string; training_id?: string; status?: string }) => {
    const q = new URLSearchParams();
    if (params?.employee_id) q.set("employee_id", params.employee_id);
    if (params?.training_id) q.set("training_id", params.training_id);
    if (params?.status) q.set("status", params.status);
    return req<TrainingAssignment[]>(`/assignments?${q}`);
  },
  completeAssignment: (id: string, d: object) =>
    req<TrainingAssignment>(`/assignments/${id}/complete`, { method: "POST", body: JSON.stringify(d) }),
  markOverdue: () => req<{ marked_overdue: number }>("/assignments/mark-overdue", { method: "POST" }),

  // Certifications
  createCertification: (d: object) =>
    req<CertificationRecord>("/certifications", { method: "POST", body: JSON.stringify(d) }),
  listCertifications: (params?: { employee_id?: string; status?: string; expiring_days?: number }) => {
    const q = new URLSearchParams();
    if (params?.employee_id) q.set("employee_id", params.employee_id);
    if (params?.status) q.set("status", params.status);
    if (params?.expiring_days != null) q.set("expiring_days", String(params.expiring_days));
    return req<CertificationRecord[]>(`/certifications?${q}`);
  },
  getCertification: (id: string) => req<CertificationRecord>(`/certifications/${id}`),
  updateCertification: (id: string, d: object) =>
    req<CertificationRecord>(`/certifications/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
  refreshCertStatuses: () => req<{ updated: number }>("/certifications/refresh-statuses", { method: "POST" }),

  // Feedback
  submitFeedback: (d: object) => req<TrainingFeedback>("/feedback", { method: "POST", body: JSON.stringify(d) }),
  listFeedback: (training_id?: string) =>
    req<TrainingFeedback[]>(`/feedback${training_id ? `?training_id=${training_id}` : ""}`),

  // Reports
  reportCompletion: (training_id?: string) =>
    req<object>(`/reports/completion${training_id ? `?training_id=${training_id}` : ""}`),
  reportCertExpiry: (days?: number) =>
    req<object[]>(`/reports/certification-expiry${days ? `?days=${days}` : ""}`),
  reportTrainingCost: () => req<object[]>("/reports/training-cost"),
  reportSkillGaps: (department_id?: string) =>
    req<EmployeeSkillRow[]>(`/reports/skill-gaps${department_id ? `?department_id=${department_id}` : ""}`),

  // AI
  runSkillGapAdvisor: () => req<TRAIRec[]>("/ai/run-skill-gap-advisor", { method: "POST" }),
  runEffectivenessAnalyzer: () => req<TRAIRec[]>("/ai/run-effectiveness-analyzer", { method: "POST" }),
  runComplianceRisk: () => req<TRAIRec[]>("/ai/run-compliance-risk", { method: "POST" }),
  listAIRecs: (status?: string) =>
    req<TRAIRec[]>(`/ai/recommendations${status ? `?status=${status}` : ""}`),
  ackAIRec: (id: string, d: object) =>
    req<TRAIRec>(`/ai/recommendations/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
};

File: backend/app/models/training.py
from __future__ import annotations
import uuid
from datetime import datetime, date
from decimal import Decimal
from enum import Enum as PyEnum
from sqlalchemy import (
    Column, String, Numeric, Boolean, DateTime, Date,
    ForeignKey, Text, Integer, Enum, UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base


class SkillCategory(str, PyEnum):
    TECHNICAL = "technical"
    SOFT = "soft"
    SAFETY = "safety"
    COMPLIANCE = "compliance"
    MANAGEMENT = "management"


class SkillLevel(str, PyEnum):
    BASIC = "basic"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"
    EXPERT = "expert"


class TrainingType(str, PyEnum):
    INTERNAL = "internal"
    EXTERNAL = "external"
    ONLINE = "online"
    CLASSROOM = "classroom"
    ON_JOB = "on_job"


class SessionStatus(str, PyEnum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class AssignmentStatus(str, PyEnum):
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    OVERDUE = "overdue"
    WAIVED = "waived"


class CertificationStatus(str, PyEnum):
    VALID = "valid"
    EXPIRING = "expiring"
    EXPIRED = "expired"
    REVOKED = "revoked"


class TRAIAgentType(str, PyEnum):
    SKILL_GAP_ADVISOR = "skill_gap_advisor"
    EFFECTIVENESS_ANALYZER = "effectiveness_analyzer"
    COMPLIANCE_RISK = "compliance_risk"


class TRAIRecStatus(str, PyEnum):
    PENDING = "pending"
    ACKNOWLEDGED = "acknowledged"
    ACTIONED = "actioned"
    DISMISSED = "dismissed"


class SkillMaster(Base):
    __tablename__ = "training_skills"

    skill_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    skill_code = Column(String(50), unique=True, nullable=False)
    skill_name = Column(String(200), nullable=False)
    category = Column(Enum(SkillCategory), nullable=False)
    description = Column(Text, nullable=True)
    active_flag = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    employee_profiles = relationship("EmployeeSkillProfile", back_populates="skill", lazy="dynamic")


class EmployeeSkillProfile(Base):
    __tablename__ = "employee_skill_profiles"

    employee_skill_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(String(100), nullable=False)
    employee_name = Column(String(200), nullable=True)
    department_id = Column(String(100), nullable=True)
    skill_id = Column(UUID(as_uuid=True), ForeignKey("training_skills.skill_id"), nullable=False)
    current_level = Column(Enum(SkillLevel, name="training_skilllevel"), nullable=True)
    required_level = Column(Enum(SkillLevel, name="training_skilllevel"), nullable=True)
    last_assessed_date = Column(Date, nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    skill = relationship("SkillMaster", back_populates="employee_profiles")

    __table_args__ = (
        UniqueConstraint("employee_id", "skill_id", name="uq_employee_skill"),
    )


class TrainingProgram(Base):
    __tablename__ = "training_programs"

    training_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    training_code = Column(String(50), unique=True, nullable=False)
    training_name = Column(String(300), nullable=False)
    training_type = Column(Enum(TrainingType), nullable=False)
    category = Column(Enum(SkillCategory), nullable=False)
    duration_hours = Column(Numeric(6, 2), nullable=True)
    provider = Column(String(200), nullable=True)
    cost = Column(Numeric(12, 2), nullable=True)
    mandatory_flag = Column(Boolean, default=False)
    validity_period_days = Column(Integer, nullable=True)
    active_flag = Column(Boolean, default=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    sessions = relationship("TrainingSession", back_populates="program", lazy="dynamic")
    assignments = relationship("TrainingAssignment", back_populates="program", lazy="dynamic")
    certifications = relationship("CertificationRecord", back_populates="program", lazy="dynamic")


class TrainingSession(Base):
    __tablename__ = "training_sessions"

    session_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    training_id = Column(UUID(as_uuid=True), ForeignKey("training_programs.training_id"), nullable=False)
    session_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=True)
    location = Column(String(300), nullable=True)
    trainer_id = Column(String(100), nullable=True)
    trainer_name = Column(String(200), nullable=True)
    capacity = Column(Integer, nullable=True)
    enrolled_count = Column(Integer, default=0)
    status = Column(Enum(SessionStatus), default=SessionStatus.PLANNED, nullable=False)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    program = relationship("TrainingProgram", back_populates="sessions")
    assignments = relationship("TrainingAssignment", back_populates="session", lazy="dynamic")


class TrainingAssignment(Base):
    __tablename__ = "training_assignments"

    assignment_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(String(100), nullable=False)
    employee_name = Column(String(200), nullable=True)
    department_id = Column(String(100), nullable=True)
    training_id = Column(UUID(as_uuid=True), ForeignKey("training_programs.training_id"), nullable=False)
    session_id = Column(UUID(as_uuid=True), ForeignKey("training_sessions.session_id"), nullable=True)
    assigned_by = Column(String(200), nullable=True)
    due_date = Column(Date, nullable=True)
    completion_date = Column(Date, nullable=True)
    status = Column(Enum(AssignmentStatus), default=AssignmentStatus.ASSIGNED, nullable=False)
    score = Column(Numeric(5, 2), nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    program = relationship("TrainingProgram", back_populates="assignments")
    session = relationship("TrainingSession", back_populates="assignments")

    __table_args__ = (
        UniqueConstraint("employee_id", "training_id", "session_id", name="uq_assignment_employee_training_session"),
    )


class CertificationRecord(Base):
    __tablename__ = "certification_records"

    certification_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(String(100), nullable=False)
    employee_name = Column(String(200), nullable=True)
    training_id = Column(UUID(as_uuid=True), ForeignKey("training_programs.training_id"), nullable=True)
    certificate_name = Column(String(300), nullable=False)
    certificate_no = Column(String(100), nullable=True)
    issuer = Column(String(200), nullable=True)
    issue_date = Column(Date, nullable=False)
    expiry_date = Column(Date, nullable=True)
    status = Column(Enum(CertificationStatus), default=CertificationStatus.VALID, nullable=False)
    document_url = Column(String(500), nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    program = relationship("TrainingProgram", back_populates="certifications")


class TrainingFeedback(Base):
    __tablename__ = "training_feedback"

    feedback_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(String(100), nullable=False)
    employee_name = Column(String(200), nullable=True)
    training_id = Column(UUID(as_uuid=True), ForeignKey("training_programs.training_id"), nullable=False)
    session_id = Column(UUID(as_uuid=True), ForeignKey("training_sessions.session_id"), nullable=True)
    rating = Column(Integer, nullable=False)
    content_rating = Column(Integer, nullable=True)
    trainer_rating = Column(Integer, nullable=True)
    relevance_rating = Column(Integer, nullable=True)
    comments = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    __table_args__ = (
        UniqueConstraint("employee_id", "training_id", "session_id", name="uq_feedback_employee_training_session"),
    )


class TRAIRecommendation(Base):
    __tablename__ = "tr_ai_recommendations"

    rec_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_type = Column(Enum(TRAIAgentType), nullable=False)
    status = Column(Enum(TRAIRecStatus), default=TRAIRecStatus.PENDING, nullable=False)
    title = Column(String(300), nullable=False)
    body = Column(Text, nullable=False)
    employee_id = Column(String(100), nullable=True)
    department_id = Column(String(100), nullable=True)
    training_id = Column(UUID(as_uuid=True), nullable=True)
    certification_id = Column(UUID(as_uuid=True), nullable=True)
    score = Column(Integer, nullable=True)
    actioned_by = Column(String(200), nullable=True)
    actioned_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
